### PR TITLE
Improve model accuracy by using F32 P*V with v_cache dot product

### PR DESCRIPTION
**Pull request by guoqingbao**
**Use higher precision data type (F32) for P*V with v_cache dot product**

## Purpose

The existing paged attention kernel uses f16/bf16 (input data type) for computing P*V and v_cache dot products. This can sacrifice accuracy for certain models (including the Qwen3 series), particularly when dealing with long contexts, longer outputs, and quantized models.

This PR switches to F32 for the P*V with v_cache dot product, which significantly improves model accuracy.

## Test Plan

Because the full vLLM project is large, we validated the change using a migrated Rust project ([candle-vllm](https://github.com/EricLBuehler/candle-vllm)) that uses identical paged attention kernels with a simpler project structure.

Reproducible steps:

1. Clone and checkout repo before this PR (e.g. `git checkout f6b99f2`), then build candle-vllm without the `paged-attn` feature.
2. Run the Qwen3-MoE model:
```shell
   cargo run --features cuda,nccl --release -- --w /data/Qwen3-30B-A3B-Instruct-2507 --isq q4k --d 0,1
````

where `/data/Qwen3-30B-A3B-Instruct-2507` is the local model path.

3. Start the simple chatbot:

```shell
python3 examples/chat.py
```

Ask multi-turn questions and test scenarios requiring the model to maintain context across long outputs (e.g., >16k tokens). The current kernel often loses focus or fails to follow instructions.

4. Pull the latest commit (with this PR), rebuild candle-vllm, and re-run with the same command and inputs. Accuracy improves significantly.

## Test Results

* Applying this PR substantially improves Qwen3-30B-A3B accuracy on A100 GPUs (and other GPUs in theory).
* Other models are also expected to benefit from the higher precision.

There is another test on V100 which shows promising improvements by leveraging this PR:

https://github.com/guoqingbao/vllm.rs/pull/28

## Side Effects

We did not observe obvious decoding speed degradation, here's explain why?

* In the original code, `logits` were downcast from F32 to f16/bf16 `logits_vec`:

  ```c++
  // L_vec logits_vec;
  // from_float(logits_vec, *reinterpret_cast<Float_L_vec*>(logits + token_idx - start_token_idx));
  ```

  In the proposed code, we directly use the F32 logits:

  ```c++
  Float_L_vec logits_vec = *reinterpret_cast<Float_L_vec*>(logits + token_idx - start_token_idx);
  ```

  → This avoids unnecessary type conversion.

* For the dot product, the old code used f16/bf16:

  ```c++
  // accs[i] += dot(logits_vec, v_vec);
  ```

  Now we upcast `v_vec` to F32 before the dot product (minimal cost):

  ```c++
  Float_L_vec v_vec_f32 = to_float(v_vec);
  accs[i] += dot(logits_vec, v_vec_f32);
  ```

  → While F32 dot products may have slightly higher cost, GPU execution leverages both fp16/bf16 and F32 compute units in parallel, so we did not observe a slowdown.

---

<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

* [x] Fix accuracy degradation for paged attention kernel.
* [x] Validate with Rust version of vLLM for simplicity.
* [x] Provide reproducible steps and e2e test results.
* [ ] (Optional) Update documentation (e.g., `supported_models.md`, `examples`).
* [ ] (Optional) Update release notes if user-facing.

</details>

### CODE REVIEW

**Comment by sempervictus** [comment]
have been testing iterations of attention work by @guoqingbao in `candle-vllm` and `vllm.rs` on SM70 and 89 hardware and this is "night and day" for accuracy on contexts exceeding ~20k in pre-BF16 land. NV might be killing the old gear w/ CUDA13 but people still have it and this does make their life a lot better when FA is not an option.
